### ISSUE TITLE
'Expected expression' error in opencl tests due to undefined TEST_DIR

### ISSUE THREAD

**Issue by schuhschuh** [arrayfire/arrayfire#1494]
When building with `BUILD_TEST` option `ON` in Xcode, I get multiple `Expected expression` errors in lines which use the `TEST_DIR` macro. Detailed error message for `moments_opencl`, and similar errors for `harris_opencl`:

```
CompileC Xcode/test/ARRAYFIRE.build/Release/moments_opencl.build/Objects-normal/x86_64/moments.o test/moments.cpp normal x86_64 c++ com.apple.compilers.llvm.clang.1_0.compiler
    cd /Users/as12312/Software/ArrayFire
    export LANG=en_US.US-ASCII
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -x c++ -arch x86_64 -fmessage-length=0 -fdiagnostics-show-note-include-stack -fmacro-backtrace-limit=0 -Wno-trigraphs -fpascal-strings -O3 -Wno-missing-field-initializers -Wno-missing-prototypes -Wno-return-type -Wno-non-virtual-dtor -Wno-overloaded-virtual -Wno-exit-time-destructors -Wno-missing-braces -Wparentheses -Wswitch -Wno-unused-function -Wno-unused-label -Wno-unused-parameter -Wno-unused-variable -Wunused-value -Wno-empty-body -Wno-uninitialized -Wno-unknown-pragmas -Wno-shadow -Wno-four-char-constants -Wno-conversion -Wno-constant-conversion -Wno-int-conversion -Wno-bool-conversion -Wno-enum-conversion -Wno-shorten-64-to-32 -Wno-newline-eof -Wno-c++11-extensions -DCMAKE_INTDIR=\"Release\" -DWITH_FREEIMAGE -DGLEW_MX -DWITH_GRAPHICS -DOS_MAC -DAF_UNIFIED -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk -fasm-blocks -fstrict-aliasing -Wdeprecated-declarations -Winvalid-offsetof -mmacosx-version-min=10.11 -Wno-sign-conversion -I/Users/as12312/Software/ArrayFire/Xcode/test/Release/include -I/usr/local/include -I/Users/as12312/Software/ArrayFire/Xcode/third_party/forge/include -I/usr/X11R6/include -I/Users/as12312/Software/ArrayFire/include -I/Users/as12312/Software/ArrayFire/src/backend -I/Users/as12312/Software/ArrayFire/src/api/c -I/Users/as12312/Software/ArrayFire/test/../test/gtest/include -I/Users/as12312/Software/ArrayFire/test -I/Users/as12312/Software/ArrayFire/Xcode/test/ARRAYFIRE.build/Release/moments_opencl.build/DerivedSources/x86_64 -I/Users/as12312/Software/ArrayFire/Xcode/test/ARRAYFIRE.build/Release/moments_opencl.build/DerivedSources -Wmost -Wno-four-char-constants -Wno-unknown-pragmas -F/Users/as12312/Software/ArrayFire/Xcode/test/Release -F/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks -DNDEBUG -DAF_OPENCL -Wall -fvisibility=hidden -D TEST_DIR=\\/Users/as12312/Software/ArrayFire/test/data\\ -MMD -MT dependencies -MF /Users/as12312/Software/ArrayFire/Xcode/test/ARRAYFIRE.build/Release/moments_opencl.build/Objects-normal/x86_64/moments.d --serialize-diagnostics /Users/as12312/Software/ArrayFire/Xcode/test/ARRAYFIRE.build/Release/moments_opencl.build/Objects-normal/x86_64/moments.dia -c /Users/as12312/Software/ArrayFire/test/moments.cpp -o /Users/as12312/Software/ArrayFire/Xcode/test/ARRAYFIRE.build/Release/moments_opencl.build/Objects-normal/x86_64/moments.o

/Users/as12312/Software/ArrayFire/test/moments.cpp:145:31: error: expected expression
    momentsOnImageTest(string(TEST_DIR"/moments/gray_seq_16_moments.test"), string(TEST_DIR"/imageio/gray_seq_16.png"), false);
                              ^
In file included from <built-in>:337:
<command line>:9:18: note: expanded from here
#define TEST_DIR \/Users/as12312/Software/ArrayFire/test/data\\
                 ^
/Users/as12312/Software/ArrayFire/test/moments.cpp:145:84: error: expected expression
    momentsOnImageTest(string(TEST_DIR"/moments/gray_seq_16_moments.test"), string(TEST_DIR"/imageio/gray_seq_16.png"), false);
                                                                                   ^
In file included from <built-in>:337:
<command line>:9:18: note: expanded from here
#define TEST_DIR \/Users/as12312/Software/ArrayFire/test/data\\
                 ^
/Users/as12312/Software/ArrayFire/test/moments.cpp:150:31: error: expected expression
    momentsTest<float>(string(TEST_DIR"/moments/simple_mat_batch_moments.test"));
                              ^
In file included from <built-in>:337:
<command line>:9:18: note: expanded from here
#define TEST_DIR \/Users/as12312/Software/ArrayFire/test/data\\
                 ^
/Users/as12312/Software/ArrayFire/test/moments.cpp:155:31: error: expected expression
    momentsOnImageTest(string(TEST_DIR"/moments/color_seq_16_moments.test"), string(TEST_DIR"/imageio/color_seq_16.png"), true);
                              ^
In file included from <built-in>:337:
<command line>:9:18: note: expanded from here
#define TEST_DIR \/Users/as12312/Software/ArrayFire/test/data\\
                 ^
/Users/as12312/Software/ArrayFire/test/moments.cpp:155:85: error: expected expression
    momentsOnImageTest(string(TEST_DIR"/moments/color_seq_16_moments.test"), string(TEST_DIR"/imageio/color_seq_16.png"), true);
                                                                                    ^
In file included from <built-in>:337:
<command line>:9:18: note: expanded from here
#define TEST_DIR \/Users/as12312/Software/ArrayFire/test/data\\
                 ^
/Users/as12312/Software/ArrayFire/test/moments.cpp:160:35: error: expected expression
    momentsTest<TypeParam>(string(TEST_DIR"/moments/simple_mat_moments.test"));
                                  ^
In file included from <built-in>:337:
<command line>:9:18: note: expanded from here
#define TEST_DIR \/Users/as12312/Software/ArrayFire/test/data\\
                 ^
6 errors generated.
```

**Comment by schuhschuh** [comment]
This argument suggests rather that the `TEST_DIR` is not defined properly:
`-D TEST_DIR=\\/Users/as12312/Software/ArrayFire/test/data\\`

**Comment by schuhschuh** [comment]
And here the CMake code that causes the issue in `test/CMakeLists.txt:134`:

```
IF (${CMAKE_GENERATOR} STREQUAL "Xcode")
    ADD_DEFINITIONS("-D TEST_DIR=\"\\\\\"${TESTDATA_SOURCE_DIR}\\\\\"\"")
ELSE (${CMAKE_GENERATOR} STREQUAL "Xcode")
    ADD_DEFINITIONS("-D TEST_DIR=\"\\\"${TESTDATA_SOURCE_DIR}\\\"\"")
ENDIF (${CMAKE_GENERATOR} STREQUAL "Xcode")
```

Preparing a PR...

**Comment by pavanky** [comment]
There should probably be a cleaner way to do this than having special cases for generators.

@9prady9 any idea why we have this sequence of slashes ?

**Comment by schuhschuh** [comment]
Looking at it, it seems it should be the same for any generator, i.e., the else branch. Trying it with my version of CMake (3.5.2) and Xcode (7.3.1). Possibly this was a fix for an issue with earlier versions of either of these... ?

**Comment by pavanky** [comment]
@schuhschuh Can we wait until @9prady9 responds ? He is in a different time zone so it may be a while before he gets a chance to look at it :)

**Comment by 9prady9** [comment]
@pavanky  @schuhschuh 
This additional conditional statement was added about an year ago for `Xcode` generators as they needed extra escape characters to appropriately define the path via macro. With the latest versions (either Xcode or CMake), this may not be an issue any more. I will have to check with which version the issue got fixed.

**Comment by schuhschuh** [comment]
See [release notes of CMake 3.5.0-rc3](https://cmake.org/pipermail/cmake-developers/2016-February/027841.html):

* The "Xcode" generator was fixed to escape backslashes in strings
  consistently with other generators.  Projects that previously worked
  around the inconsistecy with an extra level of backslashes
  conditioned on the Xcode generator must be updated to remove the
  workaround for CMake 3.5 and greater.

**Comment by schuhschuh** [comment]
Shall I submit a PR with the fix for CMake >= 3.5 ?

**Comment by 9prady9** [comment]
@schuhschuh Sure, please send it in. Thank you for looking into it.

**Comment by schuhschuh** [comment]
@9prady9 OK, just noticed the message. I'll remove the other PR then.

**Comment by 9prady9** [comment]
@schuhschuh That's fine, leave it as it is now. We will take care of it.

**Comment by 9prady9** [comment]
Never mind, just saw that you already closed #1499
